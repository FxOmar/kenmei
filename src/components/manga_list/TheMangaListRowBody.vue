<template lang="pug">
  .row-body
    .min-w-0.flex-1.px-4.lg_grid.lg_grid-cols-8.lg_gap-4
      .relative.col-span-3.lg_flex.lg_min-w-0.lg_items-center
        a.text-sm.font-medium.leading-5.text-gray-700.select-auto.hover_text-blue-600.lg_truncate(
          ref="titleLink"
          :href="item.links.series_url"
          :content="title"
          target="_blank"
          rel="noreferrer"
          v-text="title"
          v-tippy="tippyConfig"
        )
      .flex.col-span-1.mt-2.sm_mt-0
        .text-sm.font-normal.leading-5.text-gray-900.items-center.flex
          base-badge(:type="entryType")
            | {{ entryStatusName }}
      .col-span-2.inline-block.mt-2.mr-1.sm_m-0.sm_block
        chapter(
          :chapter="lastChapterRead"
          :volume="lastVolumeRead"
          :url="item.links.last_chapter_read_url"
        )
        .mt-2.flex.items-center.text-sm.leading-5.text-gray-500(v-if="$screen.lg")
          icon-calendar.flex-shrink-0.h-5.w-5.text-gray-400(class='mr-1.5')
          span
            time(
              v-if="item.attributes.last_read_at"
              :title='formatTime(item.attributes.last_read_at)'
              :datetime='item.attributes.last_read_at'
            )
              | Read {{ item.attributes.last_read_at | timeAgo }}
            template(v-else-if='item.attributes.last_chapter_read')
              | Last read is unknown
            template(v-else)
              | Haven't read yet
      .col-span-2.inline-block.sm_block
        chapter(
          :chapter="lastChapterAvailable"
          :volume="lastVolumeAvailable"
          :url="item.links.last_chapter_available_url"
          type='released'
        )
        .mt-2.flex.items-center.text-sm.leading-5.text-gray-500(v-if="$screen.lg")
          icon-calendar.flex-shrink-0.h-5.w-5.text-gray-400(class='mr-1.5')
          span
            time(
              v-if="item.attributes.last_released_at"
              :title='formatTime(item.attributes.last_released_at)'
              :datetime='item.attributes.last_released_at'
            )
              | Released {{ item.attributes.last_released_at | timeAgo }}
            template(v-else)
              | Last Released is unknown
</template>

<script>
  import { mapState } from 'vuex';
  import he from 'he';
  import dayjs from 'dayjs';
  import relativeTime from 'dayjs/plugin/relativeTime';
  import localizedFormat from 'dayjs/plugin/localizedFormat';

  import Chapter from './TheMangaListRowBodyChapter.vue';

  dayjs.extend(relativeTime);
  dayjs.extend(localizedFormat);

  export default {
    components: {
      Chapter,
    },
    filters: {
      timeAgo(datetime) {
        const relative = dayjs().to(dayjs(datetime));

        // TODO: I should evalute if I can update configuration directly
        // https://day.js.org/docs/en/customization/relative-time
        if (['in a few seconds', 'a few seconds ago'].includes(relative)) {
          return 'just now';
        }

        return relative;
      },
    },
    props: {
      item: {
        type: Object,
        required: true,
      },
    },
    data() {
      return {
        tippyConfig: {
          placement: 'top-center',
          interactive: true,
          allowHTML: false,
          onShow: () => this.tippyEnabled,
        },
      };
    },
    computed: {
      ...mapState('lists', [
        'statuses',
      ]),
      title() {
        return he.decode(this.item.attributes.title);
      },
      lastVolumeRead() {
        return this.item.attributes.last_volume_read;
      },
      lastChapterRead() {
        return this.item.attributes.last_chapter_read;
      },
      lastVolumeAvailable() {
        return this.item.attributes.last_volume_available;
      },
      lastChapterAvailable() {
        return this.item.attributes.last_chapter_available;
      },
      entryStatusName() {
        return this.statuses.find(
          (s) => s.enum === this.item.attributes.status,
        ).name;
      },
      entryType() {
        const { status } = this.item.attributes;

        return {
          1: 'success', 2: 'warning', 3: 'warning-light', 5: 'danger',
        }[status];
      },
      tippyEnabled() {
        if (!this.$screen.lg || this.$screen.touch) return;

        const { titleLink } = this.$refs;

        return titleLink.scrollWidth > titleLink.clientWidth;
      },
    },
    methods: {
      formatTime(datetime) {
        return dayjs(datetime).format('lll');
      },
    },
  };
</script>

<style lang="scss" scoped>
  .row-body {
    @apply min-w-0 flex-1 flex items-center select-none;
  }
</style>
