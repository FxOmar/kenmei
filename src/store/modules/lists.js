import { Message } from 'element-ui';
import * as userTags from '@/services/endpoints/UserTags';
import * as mangaEntries from '@/services/endpoints/v2/manga_entries';

// Can't access getter inside mutations, hence this has to be a plain function
export const getEntryIndex = (state, id) => state.entries.findIndex(
  (e) => e.id === id,
);

const state = {
  tags: [],
  entries: [],
  selectedEntries: [],
  entriesPagy: {},
  statuses: [
    { enum: 1, name: 'Reading' },
    { enum: 2, name: 'On Hold' },
    { enum: 3, name: 'Plan to Read' },
    { enum: 4, name: 'Completed' },
    { enum: 5, name: 'Dropped' },
  ],
  tagsLoading: false,
  entriesLoading: true,
  entriesUpdating: false,
  checkboxToggled: false,
};

const getters = {
  findEntryFromIDs: (state) => (ids) => state.entries.find(
    (entry) => ids.includes(entry.id),
  ),
  selectedEntriesIDs: (state) => state.selectedEntries.map((entry) => entry.id),
  selectionModeActive: (state) => state.selectedEntries.length > 0 && state.checkboxToggled,
};

const actions = {
  async getTags({ commit }) {
    const response = await userTags.index();
    const { status, data } = response;

    return status === 200 ? commit('setTags', data) : Message.error(data.error);
  },
  async getEntries({ commit }, { page, status, tagIDs, searchTerm, sort }) {
    commit('setEntriesLoading', true);

    const response = await mangaEntries.index(
      page, status, tagIDs, searchTerm, sort,
    );

    if (response.status === 200) {
      commit('setEntries', response.data.entries);
      commit('setEntriesPagy', response.data.pagy);
    } else {
      Message.error(response.data.error);
    }

    commit('setEntriesLoading', false);
  },
  async getEntriesPagy({ commit }, { page, status, tagIDs, searchTerm, sort }) {
    const response = await mangaEntries.pagyInfo(
      page, status, tagIDs, searchTerm, sort,
    );

    if (response.status === 200) {
      commit('setEntriesPagy', response.data);
    } else {
      Message.error(response.data.error);
    }
  },
};

const mutations = {
  setTags(state, data) {
    state.tags = data;
  },
  setEntries(state, data) {
    state.entries = data;
    state.selectedEntries = [];
  },
  setSelectedEntries(state, { entries, isCheckbox }) {
    state.checkboxToggled = isCheckbox;
    state.selectedEntries = entries;
  },
  setEntriesPagy(state, data) {
    state.entriesPagy = data;
  },
  addEntry(state, data) {
    state.entries.unshift(data);
  },
  addSelectedEntry(state, { entry, isCheckbox }) {
    state.checkboxToggled = isCheckbox;
    state.selectedEntries.push(entry);
  },
  updateEntry(state, data) {
    state.entries.splice(getEntryIndex(state, data.id), 1, data);
  },
  replaceEntry(state, { currentEntry, newEntry }) {
    state.entries.splice(getEntryIndex(state, currentEntry.id), 1, newEntry);
  },
  setTagsLoading(state, data) {
    state.tagsLoading = data;
  },
  setEntriesLoading(state, data) {
    state.entriesLoading = data;
  },
  setEntriesUpdating(state, data) {
    state.entriesUpdating = data;
  },
};

export default {
  namespaced: true,
  state,
  getters,
  actions,
  mutations,
};
