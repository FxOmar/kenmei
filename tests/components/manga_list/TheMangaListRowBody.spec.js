import TheMangaListRowBody from '@/components/manga_list/TheMangaListRowBody.vue';
import Chapter from '@/components/manga_list/TheMangaListRowBodyChapter.vue';

import Vuex from 'vuex';

import lists from '@/store/modules/lists';

const localVue = createLocalVue();

localVue.directive('tippy', true);
localVue.use(Vuex);

describe('TheMangaListRowBody.vue', () => {
  let store;
  let rowBody;
  let entry;

  const screen = { lg: true, touch: false };

  beforeEach(() => {
    entry = factories.entry.build();
    store = new Vuex.Store({
      modules: {
        lists: {
          namespaced: true,
          state: lists.state,
          mutations: lists.mutations,
          getters: lists.getters,
        },
      },
    });

    rowBody = shallowMount(TheMangaListRowBody, {
      store,
      localVue,
      mocks: { $screen: screen },
      propsData: { item: entry },
    });
  });

  it('shows title, status and chapters for released and last available', () => {
    expect(rowBody.html()).toMatchSnapshot();
  });

  describe('when last read is unknown', () => {
    describe('and last_chapter_read is present', () => {
      beforeEach(() => {
        entry = factories.entry.build({ attributes: { last_read_at: null } });

        rowBody.setProps({ item: entry });
      });

      it('shows Last read is unknown', () => {
        expect(rowBody.text()).toContain('Last read is unknown');
      });
    });

    describe('and last_chapter_read is not present', () => {
      beforeEach(() => {
        entry = factories.entry.build({
          attributes: { last_chapter_read: null, last_read_at: null },
        });

        rowBody.setProps({ item: entry });
      });

      it("shows haven't read yet", () => {
        expect(rowBody.text()).toContain("Haven't read yet");
      });
    });
  });

  describe('when last released is unknown', () => {
    beforeEach(() => {
      entry = factories.entry.build({ attributes: { last_released_at: null } });

      rowBody.setProps({ item: entry });
    });

    it('shows Last Released is unknown', () => {
      expect(rowBody.text()).toContain('Last Released is unknown');
    });
  });

  describe('when on mobile', () => {
    beforeEach(() => {
      rowBody = shallowMount(TheMangaListRowBody, {
        store,
        localVue,
        mocks: { $screen: { lg: false, touch: true } },
        propsData: { item: entry },
      });
    });

    it('hides timestamps', () => {
      expect(rowBody.text()).not.toContain('ago');
    });
  });
});
