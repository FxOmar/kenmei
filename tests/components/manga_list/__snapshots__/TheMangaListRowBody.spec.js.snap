// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TheMangaListRowBody.vue shows title, status and chapters for released and last available 1`] = `
<div class="row-body">
  <div class="min-w-0 flex-1 px-4 lg_grid lg_grid-cols-8 lg_gap-4">
    <div class="relative col-span-3 lg_flex lg_min-w-0 lg_items-center"><a href="example.url/manga/1" content="Manga Title" target="_blank" rel="noreferrer" class="text-sm font-medium leading-5 text-gray-700 select-auto hover_text-blue-600 lg_truncate">Manga Title</a></div>
    <div class="flex col-span-1 mt-2 sm_mt-0">
      <div class="text-sm font-normal leading-5 text-gray-900 items-center flex">
        <base-badge-stub type="success" icon="">Reading</base-badge-stub>
      </div>
    </div>
    <div class="col-span-2 inline-block mt-2 mr-1 sm_m-0 sm_block">
      <chapter-stub chapter="1" url="example.url/chapter/1" type="current"></chapter-stub>
      <div class="mt-2 flex items-center text-sm leading-5 text-gray-500">
        <icon-calendar-stub class="flex-shrink-0 h-5 w-5 text-gray-400 mr-1.5"></icon-calendar-stub><span><time title="Dec 1, 2018 12:00 AM" datetime="2018-12-01T00:00:00.000Z">Read 2 years ago</time></span>
      </div>
    </div>
    <div class="col-span-2 inline-block sm_block">
      <chapter-stub chapter="2" url="example.url/chapter/2" type="released"></chapter-stub>
      <div class="mt-2 flex items-center text-sm leading-5 text-gray-500">
        <icon-calendar-stub class="flex-shrink-0 h-5 w-5 text-gray-400 mr-1.5"></icon-calendar-stub><span><time title="Jan 1, 2019 12:00 AM" datetime="2019-01-01T00:00:00.000Z">Released 2 years ago</time></span>
      </div>
    </div>
  </div>
</div>
`;
